zrml/swaps/src/root.rs

```diff
@@ -0,0 +1,308 @@
+//     FOR ANY DAMAGES OR OTHER LIABILITY, WHETHER IN CONTRACT, TORT OR OTHERWISE,
+//     ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+//     DEALINGS IN THE SOFTWARE.
+
+use sp_runtime::traits::AtLeast32BitUnsigned;
+
+/// Calculate the preimage of `value` under `f` using bisection.
+///
+/// This function always terminates (regardless of the properties of `f`), but `f` must be
+/// a monotonous function to guarantee that the result is correct. The algorithm _assumes_ that `f`
+/// is monotonous, but we cannot ensure this. Calling with a non-monotonous function is undefined
+/// behavior. We do some sanity checks, but in the end, the consumer is responsible for providing a
+/// function which satisfies the API.
+///
+/// We return `(preimage, iteration_count)`. If `[min, max]` contains no preimage, then
+/// `min` or `max` is returned, depending on which is closer to the preimage. If `[min, max]`
+/// contains a preimage, an approximation of a preimage is returned.
+///
+/// # Arguments
+///
+/// - `f`: The function
+/// - `value`: The value of the preimage
+/// - `min`: The minimum value of the preimage
+/// - `max`: The maximum value of the preimage
+/// - `max_iterations`: Break after this many iterations
+/// - `tol`: Break if the interval is smaller than this
+pub(crate) fn calc_preimage<T, F>(
+    f: F,
+    value: T,
+    mut min: T,
+    mut max: T,
+    max_iterations: usize,
+    tol: T,
+) -> Result<(T, usize), &'static str>
+where
+    T: AtLeast32BitUnsigned + Copy,
+    F: Fn(T) -> Result<T, &'static str>,
+{
+    if min >= max {
+        return Err("calc_preimage: Function domain has no volume");
+    }
+    let fmin = f(min)?;
+    let mut fmax = f(max)?;
+    if fmin == value {
+        return Ok((min, 0));
+    } else if fmax == value {
+        return Ok((max, 0));
+    }
+
+    if is_outside_of(value, fmin, fmax) {
+        if dist(fmax, value) < dist(fmin, value) {
+            return Ok((max, 0));
+        } else {
+            return Ok((min, 0));
+        }
+    }
+    // At this point we can assume that there exists a preimage!
+
+    // Defensively use `for` instead of `while` or `loop` to ensure that it breaks after
+    // `max_iterations`.
+    let mut mid = T::zero();
+    let mut iteration_count = 0;
+    for i in 1..=max_iterations {
+        iteration_count = i;
+        let size = max.checked_sub(&min).ok_or("Unexpected arithmetic underflow")?;
+        if size < tol {
+            break;
+        }
+        mid = max.checked_add(&min).ok_or("Arithmetic overflow")? / 2u8.into();
+        let fmid = f(mid)?;
+        if fmid == value {
+            break;
+        }
+
+        // Check on which side of `value` the preimage is located.
+        if is_outside_of(value, fmid, fmax) {
+            max = mid;
+            fmax = fmid;
+        } else {
+            min = mid;
+            // We (surprisingly?) don't need this:
+            // fmin = fmid;
```

Comment:
Because `fmin` is never used before recalculation here:

`let fmid = f(mid)?;`

👍 It happens, because the old value is not relevant after the recalculation.